File: src/org/develar/mapsforgeTileServer/ChannelRegistrar.java
package org.develar.mapsforgeTileServer;

import io.netty.channel.*;
import io.netty.channel.group.ChannelGroup;
import io.netty.channel.group.DefaultChannelGroup;
import io.netty.util.concurrent.GlobalEventExecutor;
import org.jetbrains.annotations.NotNull;

@ChannelHandler.Sharable
public final class ChannelRegistrar extends ChannelInboundHandlerAdapter {
  private final ChannelGroup openChannels = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);

  public void addServerChannel(@NotNull Channel serverChannel) {
    assert serverChannel instanceof ServerChannel;
    openChannels.add(serverChannel);
  }

  @Override
  public void channelActive(ChannelHandlerContext context) throws Exception {
    // we don't need to remove channel on close - ChannelGroup do it
    openChannels.add(context.channel());

    super.channelActive(context);
  }

  public void closeAndSyncUninterruptibly() {
    openChannels.close().syncUninterruptibly();
  }
}

File: server/src/ChannelRegistrar.kt
package org.develar.mapsforgeTileServer

import io.netty.channel.*
import io.netty.channel.group.DefaultChannelGroup
import io.netty.util.concurrent.GlobalEventExecutor

ChannelHandler.Sharable
public class ChannelRegistrar() : ChannelInboundHandlerAdapter() {
  private val openChannels = DefaultChannelGroup(GlobalEventExecutor.INSTANCE)

  public fun addServerChannel(serverChannel: Channel) {
    assert(serverChannel is ServerChannel)
    openChannels.add(serverChannel)
  }

  override fun channelActive(context: ChannelHandlerContext) {
    // we don't need to remove channel on close - ChannelGroup do it
    openChannels.add(context.channel())

    super.channelActive(context)
  }

  public fun closeAndSyncUninterruptibly() {
    openChannels.close().syncUninterruptibly()
  }
}